zebra-consensus/src/block/subsidy/funding_streams.rs

```diff
@@ -0,0 +1,106 @@
+//! Funding Streams calculations. - [§7.7][7.7]
+//!
+//! [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+
+use std::convert::TryFrom;
+
+use zebra_chain::{
+    amount::{Amount, Error, NonNegative},
+    block::Height,
+    parameters::{Network, NetworkUpgrade::*},
+};
+
+use crate::{
+    block::subsidy::general::block_subsidy,
+    parameters::subsidy::{
+        FundingStreamReceiver, FUNDING_STREAM_HEIGHT_RANGES, FUNDING_STREAM_RECEIVER_DENOMINATOR,
+        FUNDING_STREAM_RECEIVER_NUMERATORS,
+    },
+};
+
+/// Returns the `fs.Value(height)` for each stream receiver
+/// as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+use std::collections::HashMap;
+pub fn funding_stream_values(
+    height: Height,
+    network: Network,
+) -> Result<HashMap<FundingStreamReceiver, Amount<NonNegative>>, Error> {
+    let canopy_height = Canopy.activation_height(network).unwrap();
+    let mut results = HashMap::new();
+
+    if height >= canopy_height {
+        let range = FUNDING_STREAM_HEIGHT_RANGES.get(&network).unwrap();
+        if range.contains(&height) {
+            for (&receiver, &numerator) in FUNDING_STREAM_RECEIVER_NUMERATORS.iter() {
+                let amount_value = (block_subsidy(height, network)?.number() as f64
+                    * (numerator as f64 / FUNDING_STREAM_RECEIVER_DENOMINATOR as f64))
+                    .floor();
+                results.insert(
+                    receiver,
+                    Amount::<NonNegative>::try_from(amount_value as i64)?,
+                );
+            }
+        }
+    }
+    Ok(results)
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Can we put these tests in a separate file?

Reply:
https://github.com/ZcashFoundation/zebra/pull/3017/commits/97d36d38989dc28a6de79f3b7d87e3a0cf0526ef